Source/utils/cel_to_cl2.cpp

```diff
@@ -0,0 +1,208 @@
+		uint32_t numFrames;
+		if (numGroups == 1) {
+			numFrames = maybeNumFrames;
+		} else {
+			numFrames = LoadLE32(data);
+			WriteLE32(&cl2Data[4 * group], cl2Data.size());
+		}
+
+		// CL2 header: frame count, frame offset for each frame, file size
+		const size_t cl2DataOffset = cl2Data.size();
+		cl2Data.resize(cl2Data.size() + 4 * (2 + static_cast<size_t>(numFrames)));
+		WriteLE32(cl2Data.data(), numFrames);
+
+		const uint8_t *srcEnd = &data[LoadLE32(&data[4])];
+		for (size_t frame = 1; frame <= numFrames; ++frame) {
+			const uint8_t *src = srcEnd;
+			srcEnd = &data[LoadLE32(&data[4 * (frame + 1)])];
+			WriteLE32(&cl2Data[cl2DataOffset + 4 * frame], static_cast<uint32_t>(cl2Data.size() - cl2DataOffset));
+
+			// Skip CEL frame header if there is one.
+			constexpr size_t CelFrameHeaderSize = 10;
+			const bool celFrameHasHeader = LoadLE16(src) == CelFrameHeaderSize;
+			if (celFrameHasHeader)
+				src += CelFrameHeaderSize;
+
+			const unsigned frameWidth = widthOrWidths.HoldsPointer() ? widthOrWidths.AsPointer()[frame - 1] : widthOrWidths.AsValue();
+
+			// Frame header: 5 16-bit offsets to 32-pixel height blocks.
+			const size_t frameHeaderPos = cl2Data.size();
+			constexpr size_t FrameHeaderSize = 10;
+			cl2Data.resize(cl2Data.size() + FrameHeaderSize);
+
+			// Frame header offset (first of five):
+			WriteLE16(&cl2Data[frameHeaderPos], FrameHeaderSize);
+
+			unsigned transparentRunWidth = 0;
+			size_t line = 0;
+			while (src != srcEnd) {
+				// Process line:
+				for (unsigned remainingCelWidth = frameWidth; remainingCelWidth != 0;) {
+					uint8_t val = *src++;
+					if (IsCelTransparent(val)) {
+						val = GetCelTransparentWidth(val);
+						transparentRunWidth += val;
+					} else {
+						AppendCl2TransparentRun(transparentRunWidth, cl2Data);
+						transparentRunWidth = 0;
+						AppendCl2PixelsOrFillRun(src, val, cl2Data);
+						src += val;
+					}
+					remainingCelWidth -= val;
+				}
+
+				// Frame header offset:
+				switch (++line) {
+				case 32:
+				case 64:
+				case 96:
+				case 128:
+					// Finish any active transparent run to not allow it to go over an offset line boundary.
+					AppendCl2TransparentRun(transparentRunWidth, cl2Data);
+					transparentRunWidth = 0;
+					WriteLE16(&cl2Data[frameHeaderPos + line / 16], static_cast<uint16_t>(cl2Data.size() - frameHeaderPos));
+					break;
+				}
+			}
+			AppendCl2TransparentRun(transparentRunWidth, cl2Data);
+		}
+
+		WriteLE32(&cl2Data[cl2DataOffset + 4 * (1 + static_cast<size_t>(numFrames))], static_cast<uint32_t>(cl2Data.size() - cl2DataOffset));
+		data = srcEnd;
+	}
+
+	auto out = std::unique_ptr<byte[]>(new byte[cl2Data.size()]);
+	memcpy(&out[0], cl2Data.data(), cl2Data.size());
+#ifdef DEBUG_CEL_TO_CL2_SIZE
+	std::cout << "\t" << size << "\t" << cl2Data.size() << "\t" << std::setprecision(1) << std::fixed << (static_cast<int>(cl2Data.size()) - static_cast<int>(size)) / ((float)size) * 100 << "%" << std::endl;

```

Comment:
A bit unusual considering we normaly use Log()?

Reply:
It's easier to copy paste without the INFO prefix (also stdout so can simply redirect to file)